src/fold/mod.rs

```diff
@@ -139,6 +140,23 @@ impl Fold for Lifetime {
     }
 }
 
+impl Fold for Substitution {
+    type Result = Substitution;
+    fn fold_with(&self, folder: &mut Folder, binders: usize) -> Result<Self::Result> {
+        let mut tys = BTreeMap::new();
+        let mut lifetimes = BTreeMap::new();
+
+        for (var, ty) in &self.tys {
+            tys.insert(*var, ty.fold_with(folder, binders)?);
```

Comment:
Maybe something like this?

```
let tys = self.tys.iter().map(|(&var, ty)| (var, ty.fold_with(folder, binders)?)).collect::<Result<BTreemap<_, _>>>()?;
```

Anyway, doesn't matter. More readable this way =)